niworkflows/interfaces/utils.py

```diff
@@ -256,3 +257,150 @@ def _gen_reference(fixed_image, moving_image, fov_mask=None, out_file=None,
+    +-------------------+------------------+------------------+------------------\
++------------------------------------------------+
+    """
+    input_spec = SanitizeImageInputSpec
+    output_spec = SanitizeImageOutputSpec
+
+    def _run_interface(self, runtime):
+        img = nb.load(self.inputs.in_file)
+        out_report = os.path.join(runtime.cwd, 'report.html')
+
+        # Retrieve xform codes
+        sform_code = int(img.header._structarr['sform_code'])
+        qform_code = int(img.header._structarr['qform_code'])
+
+        # Check qform is valid
+        valid_qform = False
+        try:
+            img.get_qform()
+            valid_qform = True
+        except ValueError:
+            pass
+
+        # Matching affines
+        matching_affines = valid_qform and np.allclose(img.get_qform(), img.get_sform())
+
+        save_file = False
+
+        # Both match, qform valid (implicit with match), codes okay -> do nothing, empty report
+        if matching_affines and qform_code > 0 and sform_code > 0:
+            self._results['out_file'] = self.inputs.in_file
+            open(out_report, 'w').close()
+
+        # Row 2:
+        elif valid_qform and qform_code > 0:
+            img.set_sform(img.get_qform(), qform_code)
+            save_file = True
+            warning_txt = 'Note on orientation: sform matrix set'
+            description = """\
+<p class="elem-desc">The sform has been copied from qform.</p>
+"""
+        # Rows 3-4:
+        # Note: if qform is not valid, matching_affines is False
+        elif sform_code > 0 and (not matching_affines or qform_code == 0):
+            img.set_qform(img.get_sform(), sform_code)
+            save_file = True
+            warning_txt = 'Note on orientation: qform matrix overwritten'
+            description = """\
+<p class="elem-desc">The qform has been copied from sform.</p>
+"""
+            if not valid_qform and qform_code > 0:
+                warning_txt = 'WARNING - Invalid qform information'
+                description = """\
+<p class="elem-desc">
+    The qform matrix found in the file header is invalid.
+    The qform has been copied from sform.
+    Checking the original qform information from the data produced
+    by the scanner is advised.
+</p>
+"""
+        # Rows 5-6:
+        else:
+            affine = img.affine
+            img.set_sform(affine, nb.nifti1.xform_codes['scanner'])
+            img.set_qform(affine, nb.nifti1.xform_codes['scanner'])
+            save_file = True
+            warning_txt = 'WARNING - Missing orientation information'
+            description = """\
+<p class="elem-desc">
+    Orientation information could not be retrieved from the image header.
+    The qform and sform matrices have been set to a default, LAS-oriented affine.
+    Analyses of this dataset MAY BE INVALID.
+</p>
+"""
+
+        if (self.inputs.force_float32 and np.dtype(img.get_data_dtype()).itemsize > 4) or self.inputs.n_volumes_to_discard:
```

Comment:
I don't fully understand the `np.dtype(img.get_data_dtype()).itemsize > 4`. So if we have `uint16` inputs and `force_float32`, we will have `uint16` outputs? That seems counterintuitive.

Reply:
Your interpretation is correct - as described in the input description. I left it this way because this is the way things have been done in the sanitizer function in MRIQC this interface is intended to replace.

Reply:
Yep, this does not make much sense. We should probably correct this (do not check `itemsize`, basically).

Reply:
The idea is that you only want to cast when the dtype is larger than 32 bit. Basically we don't want to cast int16 to float32 (which is what will happen if you remove this check). This is how MRIQC has been operating so far (see https://github.com/poldracklab/mriqc/blob/1ba709a2fa31e71e14d249b8db2fd87d5b349f53/mriqc/utils/misc.py#L55). Maybe this code can make things more clear:
```
In [14]: np.array(23,dtype=np.int16).dtype.itemsize
Out[14]: 2

In [15]: np.array(23,dtype=np.float32).dtype.itemsize
Out[15]: 4

In [16]: np.array(23,dtype=np.float64).dtype.itemsize
Out[16]: 8
```

Reply:
I see now that it's in the description, so that's good. But maybe it'd be better to rename the variable to be a little more congruent with its function.

Some options:

* `max_32bit`
* `disable_64bit`
* `allow_float64` (inverting logic)
* `allow_64bit` (inverting logic)

I don't think any of these are great, but I think they'll reduce the double-take factor.